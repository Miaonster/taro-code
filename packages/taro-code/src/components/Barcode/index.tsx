import React, {
  CSSProperties,
  forwardRef,
  useImperativeHandle,
  useMemo,
} from 'react'
import { Image } from '@tarojs/components'
import { CommonImageProps } from '../../common/types/image'
import barcode from '../../common/barcode'

export interface BarcodeProps extends CommonImageProps {
  className?: string
  text: string
  scale?: number
  width?: number
  height?: number
  style?: CSSProperties
  foregroundColor?: string
  backgroundColor?: string
}

const Barcode = forwardRef<{ image: string }, BarcodeProps>(
  (
    {
      className,
      text = '',
      scale = 4,
      width = 300,
      height = 60,
      style = {},
      foregroundColor = '#000000',
      backgroundColor = '#FFFFFF',
    },
    ref,
  ) => {
    const image = useMemo(
      () =>
        barcode({
          text,
          scale,
          whiteColor: backgroundColor,
          blackColor: foregroundColor,
        }),
      [text, scale, backgroundColor, foregroundColor],
    )
    useImperativeHandle(
      ref,
      () => {
        return { image }
      },
      [image],
    )
    const widthString = width != null ? `${width}px` : ''
    const heightString = height != null ? `${height}px` : ''
    const finalStyle = { width: widthString, height: heightString, ...style }
    return <Image className={className} style={finalStyle} src={image ?? ''} />
  },
)

export default Barcode
