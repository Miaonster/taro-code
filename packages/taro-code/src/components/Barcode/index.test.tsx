import React from 'react'
import renderer from 'react-test-renderer'
import Barcode from './index'

describe('Barcode test', () => {
  it('renders correctly with hello world', () => {
    const tree = renderer.create(<Barcode text='hello world' />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
