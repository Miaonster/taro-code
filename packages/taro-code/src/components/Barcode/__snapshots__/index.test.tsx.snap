// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Barcode test renders correctly with hello world 1`] = `
<img
  src="data:image/bmp;base64,Qk2GBwAAAAAAADYAAAAoAAAAcAIAAAEAAAABABgAAAAAAFAHAAATCwAAEwsAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////////////////////AAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////////////////////////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////////////////////AAAAAAAAAAAAAAAA////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAA////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
  style={
    {
      "height": "60px",
      "width": "300px",
    }
  }
/>
`;
